feat: reactions on Messages(sqservices -1734)

arjita-mitra opened the pull request:
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - As a Wire user I would like to have a wide range of reactions (emojis) at my disposal, so that I can use them to express myself in messages and reactions to messages.

- The **PR Description**
  - create message actions floating menu
  - integrate context menu, reply, emoji picker
  - manage action menu states focus/hover/selected
  - handle outside click
  - make everything keyboard accessible
  - unit testing
----
##### References
1. https://wearezeta.atlassian.net/browse/SQSERVICES-1734

arjita-mitra committed:
feat: create floating menu for message actions(SQSERVICES-1910)

arjita-mitra committed:
feat: manage message actions menu focus, hover, active states(SQSERVICES-1910)

- toggle message actions menu active state
- refactor message actions component and group reaction butons togehter
- add dark/light theme support
- css for action menu

arjita-mitra committed:
test: message action menu tests

arjita-mitra committed:
feat: emoji picker integration(SQSERVICES-1734)

thisisamir98 committed:
better positioning of emoji picker

arjita-mitra committed:
feat: handle outside area clicked for reaction/action menu items
overlay css
test case updated
code cleanup
context menu overlay and reset states(SQSERVICES-1734)

arjita-mitra committed:
feat: configure feature flag to enable/disable reaction

arjita-mitra committed:
Merge branch 'dev' of https://github.com/wireapp/wire-webapp into feature/SQSERVICES-1864

przemvs commented on src/script/components/MessagesList/Message/ContentMessage/ContentMessage.tsx:
```diff
@@ -185,21 +174,41 @@ const ContentMessageComponent: React.FC<ContentMessageProps> = ({
+  const isReactionFeatureEnabled = Config.getConfig().FEATURE.ENABLE_REACTION;
+
+  useEffect(() => {
+    if (isMessageFocused || msgFocusState) {
+      setActionMenuVisibility(true);
+    } else {
+      setActionMenuVisibility(false);
+    }
```
I think you can use shorter version :)
setActionMenuVisibility(isMessageFocused || msgFocusState)

arjita-mitra replied:
with if, else its more readable so keeping it this way

przemvs commented on src/script/components/MessagesList/Message/ContentMessage/MessageActions/MessageReactions.tsx:
```diff
@@ -0,0 +1,230 @@
+  const [showEmojis, setShowEmojis] = useState(false);
+  const {handleMenuOpen} = useMessageActionsState();
+  const [clientX, setPOSX] = useState(0);
+  const [clientY, setPOSY] = useState(0);
```
Move this 0 above component, e.g. const INITIAL_CLIENT_X_POS = 0;

przemvs commented on src/script/components/MessagesList/Message/ContentMessage/MessageActions/MessageReactions.tsx:
```diff
@@ -0,0 +1,230 @@
+        onKeyDown={handleMsgActionKeyDown}
+      >
+        <span aria-hidden={true}>👍</span>
+      </button>
```
Please add smth like that as props also: data-stringify-emoji=":thumbsup:"

arjita-mitra replied:
why do we need to add this prop?

arjita-mitra committed:
Merge branch 'dev' of https://github.com/wireapp/wire-webapp into feature/SQSERVICES-1864

arjita-mitra committed:
feat: add ability to react on messages(SQSERVICES-1734) (#14839)

* feat: add ability to react on messages

* feat: create util for transformReactionObj
replace thums up and heart emoji with image
image css
create list reactions component

* feat: Add emoji title utils (#14827)

* feat: convert emoji to unicode

* feat: integrate react functioanlity with reactions button
make reactions button
reactions buttons css
create reaction emoji image component

* test: unit test message reactions

* refactor: move message reactions related components under message reactions directory

* refactor: improve transformReactionObj function

call transformReactionObj from reactions list

* refactor: update test cases

* refactor: remove test case

* refactor: remove unused code
fix: review comment

* chore: update yarn lock

* chore: update yarn lock

* feat: Add ability to see messsage reactions (SQSERVICES-1876) (#14825)

* feat: Add ability to see messsage reactions (SQSERVICES-1876)

* getEmojiTitleFromEmojiUnicode

* feat: create util for transformReactionObj
replace thums up and heart emoji with image
image css
create list reactions component

* feat: Add emoji title utils (#14827)

* feat: convert emoji to unicode

* styles

* use first name

* use reduce

* use map

* use map for emoji dictionary

* remove extra ?

* remove duplicated code

* Update src/script/util/EmojiUtil.ts

Co-authored-by: Przemysław Jóźwik <przemyslaw.jozwik@wire.com>

* Update src/script/util/EmojiUtil.ts

Co-authored-by: Przemysław Jóźwik <przemyslaw.jozwik@wire.com>

* correct rendering

* yarn lock changes

* migrate styles

* dark mode handle

* change styles file name

---------

Co-authored-by: arjita-mitra <arjitamitra3@gmail.com>
Co-authored-by: Przemysław Jóźwik <przemyslaw.jozwik@wire.com>

* fix: new and older client should be in sync when a user reacts on a message

* test: add unit testing for reaction add/remove from a message

* Update src/script/components/MessagesList/Message/MessageWrapper.tsx

Co-authored-by: Thomas Belin <thomasbelin4@gmail.com>

---------

Co-authored-by: Amir Ghezelbash <thisisamir98@gmail.com>
Co-authored-by: Przemysław Jóźwik <przemyslaw.jozwik@wire.com>
Co-authored-by: Thomas Belin <thomasbelin4@gmail.com>

thisisamir98 committed:
feat: Add tooltip to reaction buttons (#14933)

* feat: Add tooltip to reaction buttons

* use isEnterKey

* onTooltipReactionCountClick

* onShowReactionDetails

* onClickReactionDetails

* remove MessageFooterLike

arjita-mitra committed:
feat: add new design for message timestamp, read receipt and integrate floating action menu inside image details modal (#14917)

* feat: refactor updateUserReactions method

* test:  updateUserReactions method tests

* feat: read receipt and timestamp new design

reaction buttons style changes

* feat: customise dark and light theme colour of emoji picker
update placeholder
remove feature flag

* fix: simplify code

* adjust position of message action menu to ensure responsivness

* feat: integrate message actions menu with detailsviewmodal

* implement new design of read receipts

* adjust style of 1 on 1 read receipts

* refactor reaction list padding to button margin to prevent extra line padding

* remove onClickReceive and replace by onClickDetails for contect menu action

* remove read status from ping message

* fix: hover, focus background colour of emoji picker

* fix: use css specificity to avoid important rule

* feat: integrate message floating action menu inside image details view modal

* test: update test cases

* fix: all typescript error fix and remove unused code

* feat: trap focus inside image modal

* feat: don't focus on reaction list buttons unless the message is focused

* feat: close emoji picker on emoji selection

* fix: message timestamp should appear after all the icons, ex: guest

* fix: place comma inline

* fix: move comment into correct place

* fix: rename method, make image detail modal id const

* fix: code review comments, update aria label

---------

Co-authored-by: V-Gira <virgile.girardin@gmail.com>

thisisamir98 committed:
runfix: Reaction views menu on 1 to 1

thisisamir98 committed:
runfix: Adjust margin of reaction tooltip

arjita-mitra committed:
chore: remove img cdn

arjita-mitra committed:
Merge branch 'dev' of https://github.com/wireapp/wire-webapp into feature/SQSERVICES-1864

arjita-mitra committed:
chore: merge commit with dev

arjita-mitra committed:
Merge branch 'feature/SQSERVICES-1864' of https://github.com/wireapp/wire-webapp into feature/SQSERVICES-1864

arjita-mitra committed:
fix: test errors

arjita-mitra committed:
fix: lock file update

arjita-mitra committed:
fix: add skiplibcheck temporarily

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/wireapp/wire-webapp?pullRequest=14771) <br/>All committers have signed the CLA.

thisisamir98 committed:
test: add withTheme

arjita-mitra committed:
fix: test cases

arjita-mitra committed:
feat: message reactions design review feedback changes(SQSERVICES-2057, SQSERVICES-2058) (#15200)

* fix: reaction improvements
increase reaction badge font size, reaction count colour update based on active/inactive, no system message background highlight

* feat: self deleting msg reactions(SQSERVICES-2057)

* Update src/script/components/MessagesList/Message/ContentMessage/MessageActions/MessageReactions/MessageReactions.styles.ts

Co-authored-by: Przemysław Jóźwik <przemyslaw.jozwik@wire.com>

* fix: code review changes

* fix: use HTMLButtonElement to reduce props passed

* fix: add missing onKeyPress attribute

---------

Co-authored-by: Przemysław Jóźwik <przemyslaw.jozwik@wire.com>

arjita-mitra committed:
fix: improve focus and hover behaviour of reactions menu (#15227)

* fix: reaction improvements
increase reaction badge font size, reaction count colour update based on active/inactive, no system message background highlight

* feat: self deleting msg reactions(SQSERVICES-2057)

* Update src/script/components/MessagesList/Message/ContentMessage/MessageActions/MessageReactions/MessageReactions.styles.ts

Co-authored-by: Przemysław Jóźwik <przemyslaw.jozwik@wire.com>

* fix: code review changes

* fix: use HTMLButtonElement to reduce props passed

* fix: add missing onKeyPress attribute

* feat: improve keyboard accessibility

remove unused code
close message action menu on tab of last interactive element of a message
prevent emoji picker up/down arrow key to naviage between messages in background
 after emoji selection/esc key then retain the focus on the emoji button

* fix: improve comments

* fix: remove array.from

---------

Co-authored-by: Przemysław Jóźwik <przemyslaw.jozwik@wire.com>

arjita-mitra committed:
fix: use emoji image inside reaction details right panel, some style updates, add reaction count (#15228)

* fix: reaction improvements
increase reaction badge font size, reaction count colour update based on active/inactive, no system message background highlight

* feat: self deleting msg reactions(SQSERVICES-2057)

* Update src/script/components/MessagesList/Message/ContentMessage/MessageActions/MessageReactions/MessageReactions.styles.ts

Co-authored-by: Przemysław Jóźwik <przemyslaw.jozwik@wire.com>

* fix: code review changes

* fix: use HTMLButtonElement to reduce props passed

* fix: add missing onKeyPress attribute

* feat: improve keyboard accessibility

remove unused code
close message action menu on tab of last interactive element of a message
prevent emoji picker up/down arrow key to naviage between messages in background
 after emoji selection/esc key then retain the focus on the emoji button

* fix: improve comments

* fix: remove array.from

* fix: replace span with emoji image for message reaction details panel

add reaction count
style change

* fix: update style name

---------

Co-authored-by: Przemysław Jóźwik <przemyslaw.jozwik@wire.com>

arjita-mitra committed:
fix: reaction QA feedback(WPB-1199) (#15321)

* fix: reaction QA feedback(WPB-1199)

- fix reaction action menu hover/focus state issues
- line break for reactions to avoid overflow

* fix: adjust action menu position on  font size change
wrap emoji pills to multi line
horizontal scroll shouldn't appear for long messages either

* fix: remove focus from reaction tooltip

* fix: add test id

* fix: [Web] Reactions: Message actions still accessible when I got removed from the conversation

* fix: context menu escape key press, multiple tooltip displayed together, stop propogating reaction action menu events

* fix: refactor code

* fix: remove unused event

* fix: remove extra div, move event handlers up to make tootlip links clickable

* chore: bump ui-kit

arjita-mitra committed:
fix: reset active menu button when no menu is open and clicked on wrapper (#15346)

tlebon committed:
chore: merge dev

tlebon committed:
chore: cleanup packages

arjita-mitra committed:
fix: remove reply message option from context menu (#15478)

arjita-mitra committed:
fix: Capitalise the first letter of the first word in the emoji name displayed on the right-side panel in the message details. (#15479)

* fix: remove reply message option from context menu

* fix: Capitalise the first letter of the first word in the emoji name displayed on the right-side panel in the message details.

* fix: reuse exisitng methods

arjita-mitra committed:
feat: emoji pills should become active/inactive when user selects/unselect an emoji[WPB-3291] (#15503)

* fix: remove reply message option from context menu

* fix: Capitalise the first letter of the first word in the emoji name displayed on the right-side panel in the message details.

* fix: reuse exisitng methods

* fix: set emoji pill active/inactive when user select/unselect a reaction

* fix: code review suggestions

arjita-mitra committed:
fix: make all emoji pills active that user ever reacted to[WPB-3291] (#15509)

* fix: remove reply message option from context menu

* fix: Capitalise the first letter of the first word in the emoji name displayed on the right-side panel in the message details.

* fix: reuse exisitng methods

* fix: set emoji pill active/inactive when user select/unselect a reaction

* fix: code review suggestions

* fix: make all emoji pills active that user ever reacted

* fix: spread the prop

arjita-mitra committed:
feat: emoji pills should be sorted by number of reactions in desc order (#15530)

* fix: remove reply message option from context menu

* fix: Capitalise the first letter of the first word in the emoji name displayed on the right-side panel in the message details.

* fix: reuse exisitng methods

* fix: set emoji pill active/inactive when user select/unselect a reaction

* fix: code review suggestions

* fix: make all emoji pills active that user ever reacted

* fix: spread the prop

* feat: sort emoji pills by reaction user count(WPB-3292)

* fix: add function description

arjita-mitra committed:
feat: retain user selected skin colour after user reacts with the selected skin colour (#15531)

* fix: remove reply message option from context menu

* fix: Capitalise the first letter of the first word in the emoji name displayed on the right-side panel in the message details.

* fix: reuse exisitng methods

* fix: set emoji pill active/inactive when user select/unselect a reaction

* fix: code review suggestions

* fix: make all emoji pills active that user ever reacted

* fix: spread the prop

* feat: sort emoji pills by reaction user count(WPB-3292)

* fix: add function description

* fix: retain user selected skin colour(WPB-3564)

* fix: add new line

arjita-mitra committed:
Merge branch 'dev' of https://github.com/wireapp/wire-webapp into feature/SQSERVICES-1864

arjita-mitra committed:
test: fix test

arjita-mitra committed:
fix: yarn lock

arjita-mitra committed:
chore: fix lock file

arjita-mitra committed:
fix: test case

atomrc commented on yalc.lock:
```diff
@@ -0,0 +1,10 @@
+{
```
This file shoundn't be commited, please remove

Commit:
fix linting

Commit:
fix: remove unecessary file

Commit:
Merge branch 'dev' into feature/SQSERVICES-1864
